task/backend/scheduler/treescheduler.go

```diff
@@ -0,0 +1,403 @@
+		sch.time = t
+		return nil
+	}
+}
+
+// Executor is any function that accepts an ID, a time, and a duration.
+func NewScheduler(executor Executor, checkpointer SchedulableService, opts ...treeSchedulerOptFunc) (*TreeScheduler, *SchedulerMetrics, error) {
+	s := &TreeScheduler{
+		executor:     executor,
+		scheduled:    btree.New(degreeBtreeScheduled),
+		nextTime:     map[ID]ordering{},
+		onErr:        func(_ context.Context, _ ID, _ time.Time, _ error) {},
+		time:         stdTime{},
+		done:         make(chan struct{}, 1),
+		checkpointer: checkpointer,
+	}
+
+	// apply options
+	for i := range opts {
+		if err := opts[i](s); err != nil {
+			return nil, nil, err
+		}
+	}
+	if s.workchans == nil {
+		s.workchans = make([]chan Item, defaultMaxWorkers)
+
+	}
+
+	s.wg.Add(len(s.workchans))
+	for i := 0; i < len(s.workchans); i++ {
+		s.workchans[i] = make(chan Item)
+		go s.work(i)
+	}
+
+	s.sm = NewSchedulerMetrics(s)
+	s.when = s.time.Now().Add(maxWaitTime)
+	s.timer = s.time.NewTimer(maxWaitTime)
+	if executor == nil {
+		return nil, nil, errors.New("Executor must be a non-nil function")
+	}
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+	schedulerLoop:
+		for {
+			select {
+			case <-s.done:
+				s.Lock()
+				s.timer.Stop()
+				// close workchans
+				for i := range s.workchans {
+					close(s.workchans[i])
+				}
+				s.Unlock()
+				return
+			case <-s.timer.C():
+				s.Lock()
+				min := s.scheduled.Min()
+				if min == nil { // grab a new item, because there could be a different item at the top of the queue
+					s.when = s.time.Now().Add(maxWaitTime)
+					s.timer.Reset(maxWaitTime) // we can reset without stop, because its fired.
+					s.Unlock()
+					continue schedulerLoop
+				}
+				it := min.(Item)
+				if it.when > s.when.UTC().Unix() {
+					s.Unlock()
+					continue schedulerLoop
+				}
+				s.process()
+				min = s.scheduled.Min()
+				if min == nil { // grab a new item, because there could be a different item at the top of the queue after processing
+					s.when = s.time.Now().Add(maxWaitTime)
+					s.timer.Reset(maxWaitTime) // we can reset without stop, because its fired.
+					s.Unlock()
+					continue schedulerLoop
+				}
+				it = min.(Item)
+				s.when = time.Unix(it.when, 0)
+				until := s.time.Until(s.when)
+				s.timer.Reset(until) // we can reset without stop, because its fired.
+				s.Unlock()
+			}
+		}
+	}()
+	return s, s.sm, nil
+}
+
+func (s *TreeScheduler) Stop() {
+	s.Lock()
+	close(s.done)
+	s.Unlock()
+	s.wg.Wait()
+}
+
+type unsent struct {
+	items []Item
+}
+
+func (u *unsent) append(i Item) {
+	u.items = append(u.items, i)
+}
+
+func (s *TreeScheduler) process() {
+	iter, toReAdd := s.iterator(s.time.Now())
+	s.scheduled.Ascend(iter)
+	for i := range toReAdd.items {
+		s.nextTime[toReAdd.items[i].id] = toReAdd.items[i].ordering
+		s.scheduled.ReplaceOrInsert(toReAdd.items[i])
+	}
+}
+
+func (s *TreeScheduler) iterator(ts time.Time) (btree.ItemIterator, *unsent) {
```

Comment:
im not sure I like the returning a pointer of unsent items. It feels like the unsent should be done as part of the iteration. Maybe you can elaborate on why this is necessary.